sdk/java-sdk-protobuf-testkit/src/main/scala/kalix/javasdk/testkit/impl/EventingTestKitImpl.scala

```diff
@@ -0,0 +1,179 @@
+import kalix.protocol.component.MetadataEntry
+import kalix.javasdk.{ Metadata => SdkMetadata }
+import kalix.javasdk.testkit.EventingTestKit.{ Message => TestKitMessage }
+
+//import kalix.javasdk.testkit.impl.EventingTestKit.EventStreamOutProbe
+//import kalix.javasdk.testkit.impl.EventingTestKit.RunningSourceProbe
+//import kalix.protocol.eventing_test_backend._
+import com.google.protobuf.ByteString
+import org.slf4j.LoggerFactory
+import scalapb.GeneratedMessage
+import akka.NotUsed
+import akka.actor.ActorSystem
+import akka.stream.BoundedSourceQueue
+import kalix.testkit.protocol.eventing_test_backend.{
+  EmitSingleCommand,
+  EventStreamOutCommand,
+  EventStreamOutElement,
+  EventStreamOutResult,
+  EventingTestKitService,
+  EventingTestKitServiceHandler,
+  RunSourceCommand,
+  SourceElem
+}
+
+import java.time.Instant
+import java.util.UUID
+import scala.concurrent.Await
+import scala.concurrent.Future
+import scala.concurrent.Promise
+import scala.concurrent.duration._
+import scala.util.Failure
+import scala.util.Success
+
+object EventingTestKitImpl {
+
+  /**
+   * Start a pretend eventing backend, configure the proxy to use this through the 'kalix.proxy.eventing.support'
+   * "grpc-backend" and the same host and port as this was started with.
+   *
+   * The returned testkit can be used to expect and emit events to the proxy as if they came from an actual pub/sub
+   * event backend.
+   */
+  def start(system: ActorSystem, host: String, port: Int): EventingTestKit = {
+
+    // Create service handlers
+    val service = new EventingTestServiceImpl(system, host, port)
+    val handler: HttpRequest => Future[HttpResponse] =
+      EventingTestKitServiceHandler(new service.ServiceImpl)(system)
+
+    val binding = Await.result(
+      Http(system)
+        .newServerAt(host, port)
+        .bind(handler),
+      10.seconds)
+    // to allow binding to 0
+    val actualPort = binding.localAddress.getPort
+    system.log.info("Eventing testkit grpc-backend started at {}:{}", host, actualPort)
+    service.port = actualPort
+    service
+  }
+
+}
+
+/**
+ * Implements the EventingTestKit protocol defined in protocols/testkit/src/main/protobuf/eventing_test_backend.proto
+ */
+private[testkit] final class EventingTestServiceImpl(system: ActorSystem, val host: String, var port: Int)
+    extends EventingTestKit {
+  import EventingTestKitImpl._
+
+  override def getHost(): String = host
+  override def getPort(): Integer = port
+
+  private val log = LoggerFactory.getLogger(classOf[EventingTestKit])
+  private implicit val sys = system
+  private implicit val ec = sys.dispatcher
+
+  private val eventDestinationProbe = new ConcurrentHashMap[EventDestination, TestProbe]()
+
+  override def getTopic(topic: String): Topic = {
+    val dest = EventDestination.defaultInstance.withTopic(topic)
+    eventDestinationProbe.putIfAbsent(dest, TestProbe())
+    val probe = eventDestinationProbe.get(dest)
```

Comment:
I think that this line could be replaced with `val probe = eventDestinationProbe.putIfAbsent(dest, TestProbe())`